Weather_daily.py

```diff
@@ -0,0 +1,122 @@
+import pandas as pd
+from abc import ABC, abstractmethod
+import openmeteo_requests
+import requests_cache
+from retry_requests import retry
+import sqlite3
+
+# Setup the Open-Meteo API client with cache and retry on error
+cache_session = requests_cache.CachedSession('.cache', expire_after=3600)
+retry_session = retry(cache_session, retries=5, backoff_factor=0.2)
+openmeteo = openmeteo_requests.Client(session=retry_session)
+
+# Data Source Class
+class DailyTemperature:
+    def __init__(self, min_temperatures: list, max_temperatures: list, dates: list):
+        self.min_temperatures = min_temperatures
+        self.max_temperatures = max_temperatures
+        self.dates = dates
+
+    def to_dataframe(self):
+        return pd.DataFrame({
+            "Date": self.dates,
+            "Min Temperature": self.min_temperatures,
+            "Max Temperature": self.max_temperatures
+        })
+
+# Abstract Service Class
+class AbstractWeatherService(ABC):
+    @abstractmethod
+    def fetch_daily_temperature(self, latitude: float, longitude: float) -> DailyTemperature:
+        pass
+
+# API Weather Service Class
+class ApiWeatherService(AbstractWeatherService):
+    def __init__(self, api_url: str):
+        self.api_url = api_url
+
+    def fetch_daily_temperature(self, latitude: float, longitude: float) -> DailyTemperature:
+        params = {
+            "latitude": latitude,
+            "longitude": longitude,
+            "daily": ["temperature_2m_max", "temperature_2m_min"],
+            "past_days": 92  # Fetch past 92 days (approximately 3 months)
+        }
+        responses = openmeteo.weather_api(self.api_url, params=params)
+        response = responses[0]
+
+        daily = response.Daily()
+        daily_temperature_2m_max = daily.Variables(0).ValuesAsNumpy()
+        daily_temperature_2m_min = daily.Variables(1).ValuesAsNumpy()
+
+        daily_data = {
+            "date": pd.date_range(
+                start=pd.to_datetime(daily.Time(), unit="s", utc=True),
+                end=pd.to_datetime(daily.TimeEnd(), unit="s", utc=True),
+                freq=pd.Timedelta(seconds=daily.Interval()),
+                inclusive="left"
+            ),
+            "temperature_2m_max": daily_temperature_2m_max,
+            "temperature_2m_min": daily_temperature_2m_min
+        }
+
+        return DailyTemperature(
+            min_temperatures=daily_temperature_2m_min,
+            max_temperatures=daily_temperature_2m_max,
+            dates=daily_data["date"]
+        )
+
+# SQLite3 Database Operations
+class WeatherDatabase:
+    def __init__(self, db_name='weather_data_storage.db'):
+        self.conn = sqlite3.connect(db_name)
+        self.create_table()
```

Comment:
For this class, you should also implement an abstract interface & use a mocked & real (SQLite) implementation

Reply:
Updated and requested for review